sky/data/storage_utils.py

```diff
@@ -250,6 +269,97 @@ def get_excluded_files(src_dir_path: str) -> List[str]:
+    - `test/*.py` - matches .py files directly in test/ directory
+    - `test/**/*.py` - matches .py files anywhere under test/ directory
+    - `**/test/**` - matches any path containing /test/ anywhere
+    - `*/test/**` - matches paths like foo/test/bar but not test/bar or
+                    a/b/test/bar
+    - `build/` - matches build directory and everything under it
+    - `./README.md` - matches README.md only at the root
+    - `.env` - matches .env file only at the root
+    - `**/.env` - matches .env files at any depth
+
+    Args:
+        abs_path: Absolute path to check
+        base_path: Base directory path to calculate relative paths from
+        patterns: List of exclusion patterns (from skyignore/gitignore)
+
+    Returns:
+        True if the path should be excluded, False otherwise
+    """
+    rel = os.path.relpath(abs_path, base_path).replace(os.path.sep, '/')
+
+    for pattern in patterns:
+        original_pattern = pattern
+        pattern = pattern.rstrip('/')
+
+        # Handle root-anchored patterns
+        if pattern.startswith('/'):
+            # Remove leading / and treat as root-anchored
+            pattern = pattern[1:]
+            path = rel
+        elif pattern.startswith('./'):
+            # For ./ prefixed patterns, match against ./ + relative path
+            path = './' + rel
+        else:
+            path = rel
+
+        # Handle trailing slash patterns (directory patterns)
+        if original_pattern.endswith('/') and '**' not in pattern:
+            # Pattern like 'build/' should match 'build' and 'build/anything'
+            if path == pattern or path.startswith(pattern + '/'):
+                return True
+            continue
+
+        # Convert glob pattern to regex more simply
+        # Escape special regex chars except * and ?
+        regex_pattern = re.escape(pattern)
+
+        # Handle ** (matches zero or more path segments)
+        # Leading ** should match from start (including empty)
+        regex_pattern = regex_pattern.replace(r'\*\*/', '(?:.*/)?')
+        # Trailing ** should match to end (including empty)
+        regex_pattern = regex_pattern.replace(r'/\*\*', '(?:/.*)?')
+        # Middle ** between slashes
+        regex_pattern = regex_pattern.replace(r'/\*\*/', '(?:/.*/)?')
+        # Standalone ** (entire pattern)
+        regex_pattern = regex_pattern.replace(r'\*\*', '.*')
+
+        # Handle * (matches any chars except /)
+        regex_pattern = regex_pattern.replace(r'\*', '[^/]*')
+
+        # Handle ? (matches single char except /)
+        regex_pattern = regex_pattern.replace(r'\?', '[^/]')
+
+        # Add anchors
+        regex_pattern = '^' + regex_pattern + '$'
+
+        # For non-** patterns with *, ensure path depth matches
+        if '*' in pattern and '**' not in pattern and '/' in pattern:
+            # Count slashes to ensure depth matches
+            if pattern.count('/') != path.count('/'):
```

Comment:
Why do we need this check at all? Seems like this should be handled by the existing substitutions (like `\*` -> `[^/]*`)